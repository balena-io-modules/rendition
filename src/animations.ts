import { css, keyframes } from 'styled-components';

const fadeIn = keyframes`
	from {
		opacity: 0;
	}

	to {
		opacity: 1;
	}
`;

const fadeOut = keyframes`
	from {
		opacity: 1;
	}

	to {
		opacity: 0;
	}
`;

export const animations = css`
	.fadeIn {
		animation: ${fadeIn};
	}

	.fadeOut {
		animation: ${fadeOut};
	}

	.animated.faster {
		animation-duration: 500ms;
	}
`;
