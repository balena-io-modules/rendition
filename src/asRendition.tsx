import difference from 'lodash/difference';
import { arrayOf, number, oneOfType, string } from 'prop-types';
import * as React from 'react';
import styled, { withTheme } from 'styled-components';
import {
	color,
	display,
	fontSize,
	height,
	maxHeight,
	maxWidth,
	minHeight,
	minWidth,
	space,
	width,
} from 'styled-system';
import { StyledSystemProps } from './common-types';
import { Tooltips } from './tooltips';
import { compose } from './utils';

// Returns the display name of a React component. Falls back to 'Component'.
export const getDisplayName = (component: React.ComponentType): string => {
	if (typeof component === 'string') {
		return component;
	}

	if (!component) {
		return 'Component';
	}

	return component.displayName || component.name || 'Component';
};

const tooltip = new Tooltips();

const prop = oneOfType([number, string, arrayOf(oneOfType([number, string]))]);

const propTypes = {
	width: prop,
	minWidth: prop,
	maxWidth: prop,
	height: prop,
	minHeight: prop,
	maxHeight: prop,
	display: prop,
	fontSize: prop,
	color: prop,
	bg: prop,
	backgroundColor: prop,
	m: prop,
	mt: prop,
	mr: prop,
	mb: prop,
	ml: prop,
	mx: prop,
	my: prop,
	p: prop,
	pt: prop,
	pr: prop,
	pb: prop,
	pl: prop,
	px: prop,
	py: prop,
};

export const styledSystemProps = Object.keys(propTypes);

const filterStyledSystemProps =
	(passthroughProps: string[]) => (Base: React.ComponentType<{ ref: any }>) => {
		const omitProps = difference(styledSystemProps, passthroughProps);
		return React.forwardRef((props: { [key: string]: any }, ref) => {
			const nextProps = { ...props };
			for (const k of omitProps) {
				delete nextProps[k];
			}
			return <Base {...nextProps} ref={ref} />;
		});
	};

export const withStyledSystem = (child: React.ComponentType) => {
	const Base = styled(child)<StyledSystemProps>`
		${space}
		${width}
    ${minWidth}
    ${maxWidth}
    ${height}
    ${minHeight}
    ${maxHeight}
    ${display}
		${fontSize}
		${color}
	`;

	Base.displayName = getDisplayName(child);

	// The styled component messes up with the propTypes typings. We don't really need typesafety here anyway, so it is safe to just ignore it.
	// @ts-expect-error
	Base.propTypes = propTypes;

	return React.forwardRef((props: any, ref: any) => {
		return <Base {...props} ref={ref} />;
	});
};

export const withTooltip = (Base: React.ComponentType) => {
	return React.forwardRef(({ ...props }: any, ref: any) => {
		if (!props.tooltip) {
			return <Base {...props} ref={ref} />;
		}
		const tooltipProps = tooltip.bindProps(props);
		delete props.tooltip;
		return props.disabled ? (
			<span
				style={{ display: 'contents' }}
				onClick={tooltipProps.onClick}
				onMouseEnter={tooltipProps.onMouseEnter}
				onMouseLeave={tooltipProps.onMouseLeave}
			>
				<Base {...props} ref={ref} />
			</span>
		) : (
			<Base {...tooltipProps} ref={ref} />
		);
	});
};

export default function asRendition<T>(
	component: React.ComponentType | React.ForwardRefExoticComponent<T>,
	additionalEnhancers: Array<
		(Base: React.ComponentType) => React.ComponentType
	> = [],
	passthroughProps: string[] = [],
) {
	return compose(
		withTheme,
		...(additionalEnhancers || []),
		withTooltip,
		withStyledSystem,
		filterStyledSystemProps(passthroughProps),
	)(component) as unknown as T;
}
