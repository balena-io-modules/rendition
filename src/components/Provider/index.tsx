import { Grommet } from 'grommet';
import merge from 'lodash/merge';
import * as React from 'react';
import styled from 'styled-components';
import { DefaultProps, Theme } from '../../common-types';
import defaultTheme from '../../theme';
import { px } from '../../utils';

const Base = styled(Grommet)`
	font-family: ${props => props.theme.font};
	font-size: ${props => px(props.theme.fontSizes[1])};

	h1,
	h2,
	h3,
	h4,
	h5,
	h6,
	button {
		font-family: ${props => props.theme.titleFont};
	}
`;

const Provider = ({ theme, ...props }: ThemedProvider) => {
	return <Base theme={merge(defaultTheme, theme)} {...props} />;
};

export interface ThemedProvider extends DefaultProps {
	theme?: Theme;
}

export default Provider;
