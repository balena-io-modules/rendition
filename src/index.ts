export { Alert, AlertProps } from './components/Alert';
export { Async } from './components/Async';
export { Badge, BadgeProps } from './components/Badge';
export { Banner, BannerProps } from './components/Banner';
export { Button, ButtonProps } from './components/Button';
export {
	ButtonGroup,
	ButtonGroupProps,
	ButtonGroupOption,
} from './components/ButtonGroup';
export { Card, CardProps } from './components/Card';
export { Checkbox, CheckboxProps } from './components/Checkbox';
export { Copy, CopyProps } from './components/Copy';
export { Divider, DividerProps } from './components/Divider';
export {
	Filters,
	FiltersProps,
	FiltersView,
	FilterRenderMode,
	FilterSignature,
} from './components/Filters';
export { Fixed, FixedProps } from './components/Fixed';
export {
	Form,
	FormProps,
	FormWidgetProps,
	RenditionUiSchema,
} from './components/Form';

export {
	Renderer,
	RendererProps,
	Widget,
	WidgetProps,
	widgetFactory,
} from './components/Renderer';

export { DataGrid, DataGridProps } from './components/DataGrid';
export {
	HighlightedName,
	HighlightedNameProps,
} from './components/HighlightedName';
export { Input, InputProps } from './components/Input';
export { Modal, ModalProps } from './components/Modal';
export {
	SurroundingOverlay,
	SurroundingOverlayProps,
	PartialDomRect,
	getBoundingContainerRect,
} from './components/SurroundingOverlay';
export { Pager, PagerProps } from './components/Pager';
export { ProgressBar, ProgressBarProps } from './components/ProgressBar';
export { Provider } from './components/Provider';
export { Rating, RatingProps } from './components/Rating';
export { Search, SearchProps } from './components/Search';
export { Select, SelectProps } from './components/Select';
export { Sidebar, SidebarProps } from './components/Sidebar';
export { Spinner, SpinnerProps } from './components/Spinner';
export {
	StatsBar,
	StatsBarProps,
	StatsTitle,
	StatsTitleProps,
	ValueWithMaxTitle,
	ValueWithMaxTitleProps,
} from './components/StatsBar';
export { Steps, Step, StepsProps } from './components/Steps';
export { RadioButton, RadioButtonProps } from './components/RadioButton';
export {
	RadioButtonGroup,
	RadioButtonGroupProps,
} from './components/RadioButtonGroup';

export {
	Table,
	TableRow,
	TableColumn,
	TableSortOptions,
	TableProps,
} from './components/Table';
export { Accordion, AccordionProps } from './components/Accordion';
export { Terminal, TerminalProps } from './components/Terminal';
export { Txt, TxtProps } from './components/Txt';
export { Textarea, TextareaProps } from './components/Textarea';
export { Img, ImgProps } from './components/Img';
export { Heading } from './components/Heading';
export {
	NotificationsContainer,
	notifications,
} from './components/Notifications';
export { Container } from './components/Container';
export {
	DropDownButton,
	DropDownButtonProps,
	DropdownOption,
} from './components/DropDownButton';
export { Map, MapProps } from './components/Map';
export { Navbar, NavbarProps } from './components/Navbar';
export { Link, LinkProps } from './components/Link';
export { List, ListProps } from './components/List';
export { Popover, PopoverProps, PopoverOptions } from './components/Popover';
export { Tab, TabProps, Tabs, TabsProps } from './components/Tabs';
export { Tag, TagProps, TagItem } from './components/Tag';
export {
	TagManagementModal,
	TagManagementModalProps,
} from './components/TagManagementModal';
export { CollectionSummary } from './components/TagManagementModal/CollectionSummary/CollectionSummary';
export { CaretButton } from './components/TagManagementModal/CollectionSummary/CaretButton';
export { Avatar, AvatarProps } from './components/Avatar';
export { Box, BoxProps } from './components/Box';
export { Breadcrumbs, BreadcrumbsProps, Crumb } from './components/Breadcrumbs';
export { Flex, FlexProps } from './components/Flex';
export { useTheme } from './hooks/useTheme';
export { useBreakpoint } from './hooks/useBreakpoint';
export { useRequest } from './hooks/useRequest';
export { createPoll, Poll } from './utils/poll';
export { default as Theme } from './theme';
export { Theme as ThemeType } from './common-types';

export {
	default as asRendition,
	withStyledSystem,
	withTooltip,
} from './asRendition';

import * as SchemaSieve from './components/Filters/SchemaSieve';
export { SchemaSieve };

export {
	AutoUI,
	AutoUIProps,
	autoUIRunTransformers,
	autoUIDefaultPermissions,
	autoUIGetModelForCollection,
	autoUIAddToSchema,
	AutoUIAction,
	AutoUIBaseResource,
	AutoUIRawModel,
	AutoUIModel,
	autoUIJsonSchemaPick,
	autoUIGetDisabledReason,
} from './extra/AutoUI';
export {
	listFilterQuery,
	PersistentFilters,
} from './extra/AutoUI/Filters/PersistentFilters';
export * from './extra/AutoUI/Lenses/types';

export { AutoUIApp, AutoUIAppProps } from './extra/AutoUIApp';

export {
	ResponsiveStyle,
	StyledSystemProps,
	PineDataType,
	SchemaEntry,
	Schema,
	FilterRule,
	SingleFilterView,
	FilterViewScope,
	Coloring,
	ThemeColorSet,
	WithSemilight,
	ColorShade,
	Shading,
	Sizing,
	TooltipPlacement,
	TooltipProps,
	Tooltip,
	RenditionSystemProps,
} from './common-types';
export {
	UnstableTempDownloadImageModal,
	UnstableTempDownloadImageModalProps,
} from './unstable-temp/DownloadImageModal/DownloadImageModal';
