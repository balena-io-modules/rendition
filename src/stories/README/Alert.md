# Alert

[View story source](https://github.com/balena-io-modules/rendition/blob/master/src/stories/Alert.js)

## Props

| Name   | Type   | Default   | Required   | Description   |
| ------ | ------ | --------- | ---------- | ------------- |
| `primary`     | `boolean` | - | - | If true, use the `primary` theme color     |
| `secondary`   | `boolean` | - | - | If true, use the `secondary` theme color   |
| `tertiary`    | `boolean` | - | - | If true, use the `tertiary` theme color    |
| `quarternary` | `boolean` | - | - | If true, use the `quarternary` theme color |
| `danger`      | `boolean` | - | - | If true, use the `danger` theme color      |
| `warning`     | `boolean` | - | - | If true, use the `warning` theme color     |
| `success`     | `boolean` | - | - | If true, use the `success` theme color     |
| `info`        | `boolean` | - | - | If true, use the `info` theme color        |
| `emphasized`  | `boolean` | - | - | If true, use the use bolder colors and a larger size |
| `plaintText` | `boolean` | - | - | If true, the alert will be rendered without a border or a background |
| `prefix` | <code>JSX.Element &#124; string &#124; false</code> | - | - | Set a prefix on the alert message, if this prop is set to false, the default prefix will not be shown |
| `onDismiss` | `() => void` | - | - | A function that is called when dismissing an alert
