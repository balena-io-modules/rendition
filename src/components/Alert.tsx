import * as React from 'react';
import FaCheckCircle = require('react-icons/lib/fa/check-circle');
import FaClose = require('react-icons/lib/fa/close');
import FaExclamationCircle = require('react-icons/lib/fa/exclamation-circle');
import FaExclamationTriangle = require('react-icons/lib/fa/exclamation-triangle');
import FaInfoCircle = require('react-icons/lib/fa/info-circle');
import styled, { StyledFunction, withTheme } from 'styled-components';
import asRendition from '../asRendition';
import { bold, getColor, normal, px } from '../utils';
import { Flex } from './Grid';

interface AlertProps extends DefaultProps, Coloring, Sizing {
	plaintext?: boolean;
	prefix?: JSX.Element | string | false;
	onDismiss?: () => void;
}

const getPadding = (props: AlertProps) =>
	props.emphasized ? '15px 40px' : '8px 32px';

const getTitle = (props: AlertProps) => {
	if (props.prefix === false) {
		return;
	}
	if (props.prefix !== undefined) {
		return props.prefix;
	}
	return props.danger
		? 'Oh no!'
		: props.warning
			? 'Warning!'
			: props.success
				? 'Great!'
				: props.info
					? 'Hey!'
					: '';
};

const AlertTitle = styled.span`
	display: inline-block;
	margin-right: 8px;
`;

// Firefox didn't middle align absolute positioned elements
// using flex, so we had to use an extra wrapper element
const DismissButtonWrapper = (styled.div as StyledFunction<AlertProps>)`
	display: inline-flex;
	align-items: center;
	justify-content: center;
	position: absolute;
	top: 0;
	bottom: 0;
	right: ${props => px(props.emphasized ? 20 : 12)}
`;

const DismissButton = styled.button`
	padding: 0;
	border: 0;
	background: none;
	line-height: 0;
	font-size: 14px;
	cursor: pointer;

	&:hover {
		color: black;
	}
`;

const AlertBase = (styled.div as StyledFunction<AlertProps>)`
	display: flex;
	align-items: center;
	justify-content: normal;

	position: relative;
	min-height: ${props =>
		px(props.emphasized ? props.theme.space[5] : props.theme.space[4])};
	padding: ${props => getPadding(props)};
	margin: 0;
	border-radius: ${props => px(props.theme.radius)};
	font-family: inherit;
	font-size: 16px;
	font-kerning: none;
	font-weight: ${props => normal(props)};
	appearance: none;
	text-decoration: none;
	vertical-align: middle;
	line-height: 1.1;
`;

// That's the normal alert
const Outline = AlertBase.extend`
	padding-left: 19px;
	border: 1px solid ${props => getColor(props, 'bg', 'main')};
	background: ${props => props.bg || getColor(props, 'bg', 'light')};
	color: ${props => props.theme.colors.text.main};

	& ${AlertTitle}, & ${DismissButtonWrapper} > ${DismissButton} {
		font-weight: ${props => bold(props)};
		color: ${props => getColor(props, 'color', 'main')};
	}
`;

const Filled = AlertBase.extend`
	border: 0;
	font-weight: ${props => bold(props)};
	text-align: center;
	background: ${props => props.bg || getColor(props, 'bg', 'main')};
	color: ${props => props.color || '#fff'};
`;

const Plaintext = AlertBase.extend`
	min-height: auto;
	padding: 0;
	font-size: 14px;

	&,
	& ${DismissButtonWrapper} > ${DismissButton} {
		color: ${props =>
			props.color ||
			(props.info
				? props.theme.colors.gray.main
				: getColor(props, 'color', 'main'))};
	}
`;

const getIcon = (props: AlertProps) => {
	if (props.prefix === false) {
		return;
	}
	if (props.prefix !== undefined) {
		return props.prefix;
	}
	return props.danger ? (
		<FaExclamationCircle />
	) : props.warning ? (
		<FaExclamationTriangle />
	) : props.success ? (
		<FaCheckCircle />
	) : props.info ? (
		<FaInfoCircle />
	) : (
		''
	);
};

const DismissAlert = (props: AlertProps) => (
	<DismissButtonWrapper {...props}>
		<DismissButton onClick={() => props.onDismiss && props.onDismiss()}>
			<FaClose />
		</DismissButton>
	</DismissButtonWrapper>
);

export default withTheme(
	asRendition((props: AlertProps) => {
		const { emphasized, plaintext, prefix, ...restProps } = props;
		const title = plaintext ? getIcon(props) : getTitle(props);
		if (plaintext) {
			return (
				<Plaintext {...restProps}>
					<Flex>
						{title && <AlertTitle children={title} />}
						{props.children}
					</Flex>
					{props.onDismiss && <DismissAlert {...restProps} />}
				</Plaintext>
			);
		} else if (emphasized) {
			return (
				<Filled emphasized {...restProps}>
					<div>
						{title && <AlertTitle children={title} />}
						{props.children}
					</div>
					{props.onDismiss && <DismissAlert {...restProps} />}
				</Filled>
			);
		} else {
			return (
				<Outline {...restProps}>
					<div>
						{title && <AlertTitle children={title} />}
						{props.children}
					</div>
					{props.onDismiss && <DismissAlert {...restProps} />}
				</Outline>
			);
		}
	}),
) as React.ComponentClass<AlertProps>;
