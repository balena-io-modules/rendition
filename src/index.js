import { default as Alert } from './components/Alert'
import { default as Badge } from './components/Badge'
import { default as BadgeSelect } from './components/BadgeSelect'
import { default as Banner } from './components/Banner'
import { default as Button } from './components/Button'
import { default as Card } from './components/Card'
import { default as CodeWithCopy } from './components/CodeWithCopy'
import { default as DeleteButton } from './components/DeleteButton'
import { default as DeviceStatusGauge } from './components/DeviceStatusGauge'
import { default as Divider } from './components/Divider'
import { default as Filters } from './components/Filters'
import { default as Fixed } from './components/Fixed'
import { default as Gauge } from './components/Gauge'
import { default as Input } from './components/Input'
import { default as Modal } from './components/Modal'
import { default as Pager } from './components/Pager'
import { default as Pill } from './components/Pill'
import { default as ProgressBar } from './components/ProgressBar'
import { default as Provider } from './components/Provider'
import * as SchemaSieve from './components/Filters/SchemaSieve'
import { default as Search } from './components/Search'
import { default as Select } from './components/Select'
import { default as Table } from './components/Table'
import { default as Terminal } from './components/Terminal'
import { default as TextWithCopy } from './components/TextWithCopy'
import { default as Txt } from './components/Txt'
import { default as Textarea } from './components/Textarea'
import { default as Toggle } from './components/Toggle'
import { default as Img } from './components/Img'
import { default as Heading } from './components/Heading'
import { default as Container } from './components/Container'
import { default as DropDownButton } from './components/DropDownButton'
import { default as Navbar } from './components/Navbar'
import { default as Link } from './components/Link'

import * as migrations from './migrations'
import Theme from './theme'

import { Flex, Box } from './components/Grid'

export {
  Alert,
  Badge,
  BadgeSelect,
  Banner,
  Box,
  Button,
  Card,
  CodeWithCopy,
  Container,
  DeleteButton,
  DeviceStatusGauge,
  Divider,
  DropDownButton,
  Filters,
  Fixed,
  Flex,
  Gauge,
  Heading,
  Link,
  Input,
  Img,
  Modal,
  Navbar,
  Pager,
  Pill,
  ProgressBar,
  Provider,
  SchemaSieve,
  Search,
  Select,
  Table,
  TextWithCopy,
  Txt,
  Textarea,
  Terminal,
  Theme,
  Toggle,
  migrations
}
