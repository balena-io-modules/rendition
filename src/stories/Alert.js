import * as React from 'react'
import { storiesOf } from '@storybook/react'
import { action } from '@storybook/addon-actions'
import { withScreenshot } from 'storybook-chrome-screenshot'
import withReadme from 'storybook-readme/with-readme'
import { Alert, Box, Provider } from '../'
import * as AlertReadme from './README/Alert.md'

storiesOf('Core/Alert', module)
  .addDecorator(withReadme(AlertReadme))
  .addDecorator(withScreenshot())
  .add('Standard', () => {
    return (
      <Provider>
        <Box m={3}>
          <Alert my={2} success onDismiss={action('case-3')}>
            This is a success message
          </Alert>
          <Alert my={2} danger onDismiss={action('case-1')}>
            This is an alert message
          </Alert>
          <Alert my={2} warning onDismiss={action('case-2')}>
            This is an alert message
          </Alert>
          <Alert my={2} info onDismiss={action('case-4')}>
            This is a general balena announcement
          </Alert>
        </Box>
      </Provider>
    )
  })
  .add('Emphasized', () => {
    return (
      <Provider>
        <Box m={3}>
          <Alert my={2} emphasized success onDismiss={action('case-3')}>
            This is a success message
          </Alert>
          <Alert my={2} emphasized danger onDismiss={action('case-1')}>
            This is an alert message
          </Alert>
          <Alert my={2} emphasized warning onDismiss={action('case-2')}>
            This is an alert message
          </Alert>
          <Alert my={2} emphasized info onDismiss={action('case-4')}>
            This is a general balena announcement
          </Alert>
        </Box>
      </Provider>
    )
  })
  .add('Plaintext', () => {
    return (
      <Provider>
        <Box m={3}>
          <Alert my={2} plaintext success>
            It's a success
          </Alert>
          <Alert my={2} plaintext info>
            General text goes here
          </Alert>
          <Alert my={2} plaintext danger>
            There was an error
          </Alert>
          <Alert my={2} plaintext warning>
            Warning text
          </Alert>
        </Box>
      </Provider>
    )
  })
