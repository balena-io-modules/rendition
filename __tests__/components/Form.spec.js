/* globals expect, describe, it */
import Promise from 'bluebird'
import { mount } from 'enzyme'
import React from 'react'
import renderer from 'react-test-renderer'
import sinon from 'sinon'
import Alert from '../../src/components/Alert'
import Provider from '../../src/components/Provider'
import { Form } from '../../src/unstable/'

const schema = {
  type: 'object',
  properties: {
    Name: {
      title: 'Pokemon Name',
      type: 'string'
    }
  }
}

describe('Form component', () => {
  it('should match the stored snapshot', () => {
    const component = renderer.create(
      <Provider>
        <Form schema={schema} />
      </Provider>
    )
    let tree = component.toJSON()
    expect(tree).toMatchSnapshot()
  })

  it('should still render if an unknown format is used', () => {
    const uuidSchema = {
      type: 'object',
      properties: {
        id: {
          type: 'string',
          format: 'uuid'
        }
      }
    }

    mount(
      <Provider>
        <Form schema={uuidSchema} />
      </Provider>
    )
  })

  describe('hideSubmitButton property', () => {
    it('should render a submit button if the property is not specified', () => {
      const component = mount(
        <Provider>
          <Form schema={schema} />
        </Provider>
      )

      expect(component.find('button')).toHaveLength(1)

      component.unmount()
    })

    it('should render a submit button if the property is falsey', () => {
      const component = mount(
        <Provider>
          <Form schema={schema} hideSubmitButton={false} />
        </Provider>
      )

      expect(component.find('button')).toHaveLength(1)

      component.unmount()
    })

    it('should not render a submit button if the property is truthey', () => {
      const component = mount(
        <Provider>
          <Form schema={schema} hideSubmitButton />
        </Provider>
      )

      expect(component.find('button')).toHaveLength(0)

      component.unmount()
    })
  })

  describe('submitButtonText property', () => {
    it('should change the text value of the submit button', () => {
      const submitText = 'Click me to submit this form'
      const component = mount(
        <Provider>
          <Form schema={schema} submitButtonText={submitText} />
        </Provider>
      )

      expect(component.find('button').text()).toEqual(submitText)

      component.unmount()
    })

    it('should accept a JSX element as a value', () => {
      const submitText = 'Click me to submit this form'
      const submitElement = <span>{submitText}</span>

      const component = mount(
        <Provider>
          <Form schema={schema} submitButtonText={submitElement} />
        </Provider>
      )

      expect(component.find('button').text()).toEqual(submitText)

      component.unmount()
    })
  })

  describe('submitButtonProps property', () => {
    it('should set submit button props', () => {
      const submitButtonProps = {
        className: 'custom-button-class'
      }

      const component = mount(
        <Provider>
          <Form schema={schema} submitButtonProps={submitButtonProps} />
        </Provider>
      )

      expect(component.find('button').hasClass(submitButtonProps.className)).toEqual(true)

      component.unmount()
    })
  })

  describe('onFormSubmit property', () => {
    it('should be called when the submit button is clicked', () => {
      const value = 'Squirtle'
      const callback = sinon.spy()
      const component = mount(
        <Provider>
          <Form schema={schema} onFormSubmit={callback} />
        </Provider>
      )

      const input = component.find('input')
      input.simulate('change', { target: { value } })
      component.update()

      return Promise.delay(150).then(() => {
        component.find('form').simulate('submit')
        expect(callback.callCount).toEqual(1)
        expect(callback.getCall(0).args[0].formData).toEqual({ Name: value })

        component.unmount()
      })
    })
  })

  describe('onFormChange property', () => {
    it('should be called when an input field is changed', () => {
      const value = 'Squirtle'
      const callback = sinon.spy()
      const component = mount(
        <Provider>
          <Form schema={schema} onFormChange={callback} />
        </Provider>
      )

      const input = component.find('input')
      input.simulate('change', { target: { value } })
      component.update()

      return Promise.delay(150).then(() => {
        expect(callback.callCount).toEqual(1)
        expect(callback.getCall(0).args[0].formData).toEqual({ Name: value })

        component.unmount()
      })
    })
  })

  describe('value property', () => {
    it('should set the value of the relevant input field', () => {
      const value = { Name: 'Squirtle' }
      const component = mount(
        <Provider>
          <Form schema={schema} value={value} />
        </Provider>
      )

      const input = component.find('input')
      expect(input.props().value).toEqual('Squirtle')

      component.unmount()
    })

    it('should set the value if the value is changed after the component mounts', () => {
      const value = { Name: 'Squirtle' }
      const component = mount(
        <Form schema={schema} />
      )

      component.setProps({ value })

      const input = component.find('input')
      expect(input.props().value).toEqual('Squirtle')

      component.unmount()
    })
  })

  describe('array fields', () => {
    const arraySchema = {
      type: 'object',
      properties: {
        Array: {
          type: 'array',
          items: {
            type: 'string'
          }
        }
      }
    }

    const callback = sinon.spy()
    const component = mount(
      <Provider>
        <Form
          schema={arraySchema}
          onFormChange={callback}
        />
      </Provider>
    )

    const value1 = 'foobarbaz'
    const value2 = 'bazbarfoo'

    it('should be able to add an item', () => {
      const value = 'foobarbaz'

      component.find('.rendition-form-array-item__add-item')
        .first()
        .simulate('click')

      const input = component.find('input').first()

      input.simulate('change', {
        target: {
          value: value1
        }
      })

      component.update()

      return Promise.delay(150).then(() => {
        const callArg = callback.lastCall.args[0]
        expect(callArg.formData).toEqual({
          Array: [ value ]
        })
      })
    })

    it('should be able to add multiple items', () => {
      component.find('.rendition-form-array-item__add-item')
        .first()
        .simulate('click')

      const input = component.find('input').last()

      input.simulate('change', {
        target: {
          value: value2
        }
      })

      component.update()

      return Promise.delay(150).then(() => {
        const callArg = callback.lastCall.args[0]
        expect(callArg.formData).toEqual({
          Array: [ value1, value2 ]
        })
      })
    })

    it('should be able to move an item up', () => {
      component.find('.rendition-form-array-item__move-up')
        .last()
        .simulate('click')

      component.update()

      return Promise.delay(150).then(() => {
        const callArg = callback.lastCall.args[0]
        expect(callArg.formData).toEqual({
          Array: [ value2, value1 ]
        })
      })
    })

    it('should be able to move an item down', () => {
      component.find('.rendition-form-array-item__move-down')
        .first()
        .simulate('click')

      component.update()

      return Promise.delay(150).then(() => {
        const callArg = callback.lastCall.args[0]
        expect(callArg.formData).toEqual({
          Array: [ value1, value2 ]
        })
      })
    })

    it('should be able to remove an item', () => {
      component.find('.rendition-form-array-item__remove-item')
        .last()
        .simulate('click')

      component.update()

      return Promise.delay(150).then(() => {
        const callArg = callback.lastCall.args[0]
        expect(callArg.formData).toEqual({
          Array: [ value1 ]
        })

        component.unmount()
      })
    })
  })

  describe('uiSchema property', () => {
    it('should be able to add warnings using ui:warning', () => {
      const uiSchema = {
        Name: {
          'ui:warning': 'lorem ipsum dolor sit amet'
        }
      }

      const component = mount(
        <Provider>
          <Form
            uiSchema={uiSchema}
            schema={schema}
          />
        </Provider>
      )

      const warnings = component.find(Alert)

      expect(warnings.length).toEqual(1)

      expect(warnings.first().text().trim()).toEqual(uiSchema.Name['ui:warning'])

      component.unmount()
    })
  })

  describe('field ids', () => {
    it('should generate a unique class name for each field', () => {
      const multiFieldSchema = {
        type: 'object',
        properties: {
          Name: {
            title: 'Pokemon Name',
            type: 'string'
          },
          Height: {
            type: 'number'
          },
          Weight: {
            type: 'number'
          },
          Description: {
            type: 'string'
          }
        }
      }

      const component = mount(
        <Provider>
          <Form
            schema={multiFieldSchema}
          />
        </Provider>
      )

      const expectedClassNames = Object.keys(multiFieldSchema.properties).map((key) => {
        return `rendition-form__field--root_${key}`
      })

      expectedClassNames.forEach((className) => {
        expect(component.find(`.${className}`).hostNodes()).toHaveLength(1)
      })

      component.unmount()
    })
  })
})
