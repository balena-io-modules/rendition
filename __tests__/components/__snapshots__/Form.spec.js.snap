// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Form component should match the stored snapshot 1`] = `
<div
  className="sc-kgoBCf kOSFqw"
>
  <div
    className="sc-cbkKFq jWSxbt sc-gZMcBi cmZZgb sc-dnqmqq eheJhs sc-bZQynM gBdotH"
  >
    <form
      className="rjsf"
      noValidate={false}
      onSubmit={[Function]}
    >
      <div
        className="form-group field field-object sc-gZMcBi jxxzwT sc-dnqmqq eheJhs sc-bZQynM gcuxEv"
      >
        <fieldset>
          <div
            className="form-group field field-string sc-gZMcBi jxxzwT sc-dnqmqq eheJhs sc-bZQynM gcuxEv"
          >
            <label
              className="control-label sc-eLExRp kbPUFW"
              htmlFor="root_Name"
            >
              Pokemon Name
            </label>
            <input
              autoFocus={false}
              className="sc-hmzhuo cwbGSS sc-ksYbfQ lfOsOR"
              disabled={false}
              id="root_Name"
              label="Pokemon Name"
              onBlur={[Function]}
              onChange={[Function]}
              onFocus={[Function]}
              placeholder=""
              readOnly={false}
              required={false}
              type="text"
              value=""
            />
            <div />
            <div />
          </div>
        </fieldset>
        <div />
        <div />
      </div>
      <button
        className="sc-hMqMXs kYklEl sc-kpOJdX GGSVY"
      >
        Submit
      </button>
    </form>
  </div>
</div>
`;
